# รู้จักกับ Jupyter Notebook และโมดูล `sigproc.py`

เอกสารโน้ตบุ๊คเล่มนี้จัดทำขึ้นเพื่อสาธิตการใช้งานโมดูล `sigproc.py` ในการประมวลผลและวิเคราะห์สัญญาณ ซึ่งเป็นส่วนหนึ่งของการเรียนการสอนวิชา 01204325 การสื่อสารข้อมูลและเครือข่ายคอมพิวเตอร์ ไลบรารีหลักที่นำมาใช้งานในโมดูลมีดังนี้ (ทั้งหมดมีอยู่แล้วใน Anaconda)

* [NumPy](http://www.numpy.org) สำหรับการประมวลผลเชิงแมตริกซ์
* [SciPy](http://www.scipy.org) สำหรับการคำนวณ FFT รวมถึงการอ่านและเขียนไฟล์ .wav
* [Pandas](http://pandas.pydata.org) สำหรับการรวบรวมข้อมูลในรูปตาราง
* [Bokeh](http://bokeh.pydata.org/en/latest/index.html) ในการสร้างแผนภูมิและกราฟ

## วัตถุประสงค์
1. สร้างความคุ้นเคยกับอินเตอร์เฟสของ Jupyter (IPython) Notebook
1. เข้าใจคุณสมบัติของคลื่นที่อยู่ในรูปฟังก์ชันซายน์ ซึ่งประกอบไปด้วยแอพลิจูด ความถี่ และเฟส
1. เข้าใจการมองสัญญาณในเชิงเวลา (time domain) และเชิงความถี่ (frequency domain)

## การเตรียมตัว

อิมพอร์ตโมดูล `sigproc` จากนั้นเรียกใช้ฟังก์ชัน `start_notebook()` เพื่อเริ่มต้นใช้งานดังนี้

In [2]:
from sigproc import Signal,start_notebook
bkp = start_notebook()

/usr/lib/python3.6/importlib/_bootstrap.py:219: RuntimeWarning: numpy.dtype size changed, may indicate binary incompatibility. Expected 96, got 88
  return f(*args, **kwds)
/usr/lib/python3.6/importlib/_bootstrap.py:219: RuntimeWarning: numpy.dtype size changed, may indicate binary incompatibility. Expected 96, got 88
  return f(*args, **kwds)


Loading BokehJS ...

Notebook session ready.


## สร้างสัญญาณจากการผสมคลื่นซายน์

รูปแบบทั่วไปของคลื่นซายน์สามารถเขียนได้ตามฟังก์ชัน
$$s(t) = A\cdot\sin(2\pi ft + \phi)$$
โดยที่
* $A$ คือแอพลิจูด (amplitude)
* $f$ คือความถี่ (frequency) มีหน่วยเป็นเฮิร์ตซ์ (Hz)
* $\phi$ คือเฟสชิฟต์ (phase shift) มีหน่วยเป็นเรเดียน

โค้ดด้านล่างนิยามฟังก์ชันสำหรับสร้างคลื่นซายน์ 3 รูปแบบเพื่อนำไปสร้างเป็นสัญญาณด้วยโมดูล sigproc และพล็อตเป็นกราฟเชิงเวลา

In [8]:
from numpy import sin,pi

def sine1(t):
    return 1*sin(2*pi*1*t)

def sine2(t):
    return 1*sin(2*pi*3*t+3*pi/4)

def sine3(t):
    return 1/5*sin(2*pi*5*t)

s1 = Signal(func=sine1)
s2 = Signal(func=sine2)
s3 = Signal(func=sine3)

s1.plot_time()
s2.plot_time()
s3.plot_time()

/home/ayumiizz/datacom/sigproc/sigproc.py:378: FutureWarning: Method .as_matrix will be removed in a future version. Use .values instead.
  ts.loc[interval,"a"] = func(ts[interval].t.as_matrix()-offset)


### การผสมสัญญาณ
โมดูล `sigproc` เตรียมฟังก์ชันและตัวดำเนินการที่หลากหลายเพื่อใช้ในการประมวลผลสัญญาณเบื้องต้น สัญญาณที่สร้างขึ้นมาทั้ง 3 สัญญาณสามารถนำมาผสมกันเป็นสัญญาณใหม่ได้โดยใช้ตัวดำเนินการ `+` ดังนี้

In [4]:
s = s1+s2+s3
s.plot_time(line_options=dict(color="orange",line_width=5))

เรียกใช้พร็อพเพอร์ตี `time_domain` เพื่อดึงข้อมูลของสัญญาณในรูปฟังก์ชันของเวลา ข้อมูลที่ส่งกลับมาจะอยู่ในรูปตารางที่ประกอบไปด้วยคอลัมน์ `t` แทนค่าเวลา และคอลัมน์ `a` แทนแอมพลิจูดของสัญญาณ ณ เวลาหนึ่ง ๆ

In [6]:
ts = s.time_domain
ts

t             a
0      0.000000  6.545533e-20
1      0.000045  8.548938e-04
2      0.000091  1.709787e-03
3      0.000136  2.564678e-03
4      0.000181  3.419567e-03
5      0.000227  4.274453e-03
6      0.000272  5.129334e-03
7      0.000317  5.984211e-03
8      0.000363  6.839082e-03
9      0.000408  7.693947e-03
10     0.000454  8.548804e-03
11     0.000499  9.403653e-03
12     0.000544  1.025849e-02
13     0.000590  1.111332e-02
14     0.000635  1.196814e-02
15     0.000680  1.282295e-02
16     0.000726  1.367775e-02
17     0.000771  1.453253e-02
18     0.000816  1.538730e-02
19     0.000862  1.624206e-02
20     0.000907  1.709680e-02
21     0.000952  1.795152e-02
22     0.000998  1.880623e-02
23     0.001043  1.966092e-02
24     0.001088  2.051559e-02
25     0.001134  2.137024e-02
26     0.001179  2.222487e-02
27     0.001225  2.307948e-02
28     0.001270  2.393407e-02
29     0.001315  2.478863e-02
...         ...           ...
22020  0.998685 -2.478863e-02
22021  0.998730 -2.393407e-02
22022  0.998775 -2.307948e-02
22023  0.998821 -2.222487e-02
22024  0.998866 -2.137024e-02
22025  0.998912 -2.051559e-02
22026  0.998957 -1.966092e-02
22027  0.999002 -1.880623e-02
22028  0.999048 -1.795152e-02
22029  0.999093 -1.709680e-02
22030  0.999138 -1.624206e-02
22031  0.999184 -1.538730e-02
22032  0.999229 -1.453253e-02
22033  0.999274 -1.367775e-02
22034  0.999320 -1.282295e-02
22035  0.999365 -1.196814e-02
22036  0.999410 -1.111332e-02
22037  0.999456 -1.025849e-02
22038  0.999501 -9.403653e-03
22039  0.999546 -8.548804e-03
22040  0.999592 -7.693947e-03
22041  0.999637 -6.839082e-03
22042  0.999683 -5.984211e-03
22043  0.999728 -5.129334e-03
22044  0.999773 -4.274453e-03
22045  0.999819 -3.419567e-03
22046  0.999864 -2.564678e-03
22047  0.999909 -1.709787e-03
22048  0.999955 -8.548938e-04
22049  1.000000  5.114619e-15

[22050 rows x 2 columns]

## การวิเคราะห์ความถี่ของสัญญาณ
ตามหลักการของ Fourier สัญญาณทุกสัญญาณสามารถตีความเป็นสัญญาณที่เกิดจากการนำเอาฟังก์ชันซายน์หลาย ๆ ฟังก์ชันมาผสมกันได้ ซึ่งเป็นการมองสัญญาณในเชิงความถี่ (frequency domain)

ใช้พร็อพเพอร์ตี `freq_domain` เพื่อตีความสัญญาณใน frequency domain โดยค่าที่ส่งคืนมาให้จะอยู่ในรูปตารางที่มี 3 คอลัมน์ คือ
* `f` แทนความถี่เป็นเฮิร์ตซ์
* `a` แทนแอพลิจูดสูงสุดของสัญญาณซายน์ในความถี่นั้น
* `p` แทนเฟสชิฟต์ของสัญญาณซายน์ในความถี่นั้น เทียบกับ_ฟังก์ชันโคซายน์_

In [5]:
fs = s.freq_domain
fs

f             a           p
0          0.0  1.048880e-16    0.000000
1          1.0  9.999924e-01  -89.991837
2          2.0  1.555847e-05   90.016327
3          3.0  3.333088e-01  -89.975510
4          4.0  3.572326e-05   90.032653
5          5.0  1.999482e-01  -89.959184
6          6.0  8.519434e-05   90.048980
7          7.0  5.556093e-05   90.057143
8          8.0  4.331999e-05   90.065306
9          9.0  3.612158e-05   90.073469
10        10.0  3.122488e-05   90.081633
11        11.0  2.761739e-05   90.089796
12        12.0  2.482180e-05   90.097959
13        13.0  2.257796e-05   90.106122
14        14.0  2.072966e-05   90.114286
15        15.0  1.917638e-05   90.122449
16        16.0  1.785002e-05   90.130612
17        17.0  1.670252e-05   90.138776
18        18.0  1.569884e-05   90.146939
19        19.0  1.481277e-05   90.155102
20        20.0  1.402424e-05   90.163265
21        21.0  1.331759e-05   90.171429
22        22.0  1.268042e-05   90.179592
23        23.0  1.210275e-05   90.187755
24        24.0  1.157647e-05   90.195918
25        25.0  1.109489e-05   90.204082
26        26.0  1.065246e-05   90.212245
27        27.0  1.024451e-05   90.220408
28        28.0  9.867115e-06   90.228571
29        29.0  9.516914e-06   90.236735
...        ...           ...         ...
10995  10995.0  3.877106e-08  179.755102
10996  10996.0  3.877103e-08  179.763265
10997  10997.0  3.877101e-08  179.771429
10998  10998.0  3.877099e-08  179.779592
10999  10999.0  3.877097e-08  179.787755
11000  11000.0  3.877095e-08  179.795918
11001  11001.0  3.877093e-08  179.804082
11002  11002.0  3.877091e-08  179.812245
11003  11003.0  3.877089e-08  179.820408
11004  11004.0  3.877088e-08  179.828571
11005  11005.0  3.877086e-08  179.836735
11006  11006.0  3.877084e-08  179.844898
11007  11007.0  3.877083e-08  179.853061
11008  11008.0  3.877082e-08  179.861224
11009  11009.0  3.877080e-08  179.869388
11010  11010.0  3.877079e-08  179.877551
11011  11011.0  3.877078e-08  179.885714
11012  11012.0  3.877077e-08  179.893878
11013  11013.0  3.877076e-08  179.902041
11014  11014.0  3.877075e-08  179.910204
11015  11015.0  3.877074e-08  179.918367
11016  11016.0  3.877073e-08  179.926531
11017  11017.0  3.877073e-08  179.934694
11018  11018.0  3.877072e-08  179.942857
11019  11019.0  3.877072e-08  179.951020
11020  11020.0  3.877071e-08  179.959184
11021  11021.0  3.877071e-08  179.967347
11022  11022.0  3.877071e-08  179.975510
11023  11023.0  3.877070e-08  179.983673
11024  11024.0  3.877070e-08  179.991837

[11025 rows x 3 columns]

ทดลองนำเอาข้อมูลเชิงความถี่มาพล็อตเป็นกราฟแท่งโดยใช้เมท็อด `plot_freq()` เพื่อดูแอมพลิจูดของแต่ละความถี่ และ `plot_phase()` เพื่อดูเฟสชิฟท์ 

อย่างไรก็ตามจากตารางข้างต้นจะเห็นว่าข้อมูลประกอบไปด้วยความถี่ตั้งแต่ 0 Hz ถึง 11025 Hz และค่าแอมพลิจูดของความถี่ส่วนใหญ่มีค่าน้อยมาก ๆ เราจึงจะเลือกแสดงเฉพาะความถี่ที่มีค่ามากกว่า 0.01 ดังโค้ดด้านล่าง

In [7]:
s.plot_freq(min_amplitude=0.01,fig_options=dict(height=200))
s.plot_phase(min_amplitude=0.01,fig_options=dict(height=200))

## การนำเข้าข้อมูลสัญญาณจากไฟล์ .wav
คลาส `Signal` รองรับการนำเข้าข้อมูลสัญญาณจากไฟล์ .wav โดยระบุคีย์เวิร์ดอาร์กิวเมนต์ `wav_file` ดังตัวอย่าง

In [8]:
wav = Signal(wav_file='piano-c4.wav')

ซึ่งสามารถนำข้อมูลมาวิเคราะห์และแสดงผลได้เหมือนตัวอย่างที่ผ่านมา

In [9]:
figop = dict(y_range=None)  # use automatic y-range
wav.plot_time(fig_options=figop)
wav.plot_freq(fig_options=figop)
wav.plot_phase()

## ของแถม: เล่นเสียงจากข้อมูลสัญญาณ
คลาส `Signal` เตรียมเมท็อด `play()` สำหรับนำข้อมูลสัญญาณมาเล่นเป็นข้อมูลเสียง

In [10]:
wav.play()

ตัวอย่างด้านล่างแสดงการผสมคลื่นซายน์ 3 ความถี่เพื่อให้ได้เสียงเป็นคอร์ด C

In [11]:
# สร้างสัญญาณเปล่าที่มีความยาว 2 วินาที
chord_c = Signal(duration=2)

# กำหนดความถี่และแอมพลิจูดลงในสัญญาณโดยตรง
chord_c.set_freq(262,0.3,90) # โน้ต C
chord_c.set_freq(330,0.3,90) # โน้ต E
chord_c.set_freq(392,0.3,90) # โน้ต G

# แสดงกราฟในเชิงเวลาและความถี่
chord_c.plot_time()
chord_c.plot_freq()

# เล่นสัญญาณในรูปข้อมูลเสียง
chord_c.play()